model/parallel_state.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2022 The Serverless Workflow Specification Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strconv"
+
+	validator "github.com/go-playground/validator/v10"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	val "github.com/serverlessworkflow/sdk-go/v2/validator"
+)
+
+// CompletionType define on how to complete branch execution.
+type CompletionType string
+
+const (
+	// CompletionTypeAllOf defines all branches must complete execution before the state can transition/end.
+	CompletionTypeAllOf CompletionType = "allOf"
+	// CompletionTypeAtLeast defines state can transition/end once at least the specified number of branches
+	// have completed execution.
+	CompletionTypeAtLeast CompletionType = "atLeast"
+)
+
+// ParallelState Consists of a number of states that are executed in parallel
+type ParallelState struct {
+	BaseState
+	// Branch Definitions
+	Branches []Branch `json:"branches" validate:"required,min=1,dive"`
+	// Option types on how to complete branch execution.
+	// Defaults to `allOf`
+	CompletionType CompletionType `json:"completionType,omitempty" validate:"required,oneof=allOf atLeast"`
+
+	// Used when completionType is set to 'atLeast' to specify the minimum number of branches that must complete before the state will transition."
+	// TODO: change this field to unmarshal result as int
+	NumCompleted intstr.IntOrString `json:"numCompleted,omitempty"`
```

Comment:
<img width="1013" alt="image" src="https://user-images.githubusercontent.com/6871552/195628672-5d812676-a4b8-4f14-a07b-88268ef9972e.png">

Because the specification defines the field as `number or string`. Maybe we should change the specification to simplify it.

Reply:
That's a possibility.

Reply:
Can we merge this PR first？because the implement is base on v0.8 specification.
And i will open an Issue to change the specification.